### PR TITLE
SparsityTools::distribute_sparsity_pattern with compute_index_owner

### LINKED ISSUE #8306: Replace n_..._dofs_per_processor in distribute_sparsity_pattern
In the discussion of #8298 we observed that `SparsityTools::distribute_sparsity_pattern` still uses the `n_locally_owned_dofs_per_processor` information that we want to discourage. This issue is to remind ourselves to eventually implement a better-performing alternative that avoids the MPI all-to-all communication, see also #8293.

### PR DESCRIPTION
**Pull request by kronbichler**
This PR adds alternative functions for `SparsityTools::distribute_sparsity_pattern()` and `SparsityTools::gather_sparsity_pattern()` that only take the locally owned index set of the present MPI process, rather than that of all processes. This allows us to simplify the tutorials and a few dozen of test cases, and in particular avoids the expensive call to `MPI_Allgather` discussed in #8298 and #8293, expensive meaning that with more than a few thousands of rank, the new implementation should be considerably faster. I also deprecated two of the three old functions which do not make sense any more; I left the one with the number of dofs per processor because it might be of use in some corner cases.

One of these corner cases are `dealii::Triangulation` cases where the linear system gets split among processes (but not the mesh), which I left as they were - they are mostly based off some variant of step-36.

Fixes #8306 .

This PR also significantly reduces the number of calls to `DoFHandler::compute_{n_}locally_owned_dofs_per_processor()` in the test suite. We can further eliminate a considerable number of them by initializing PETSc sparse matrices from an `IndexSet` of locally owned data rather than *all* sizes - I do not think they belong here. If we believe they are important to switch (personally I am indifferent), we can open an issue to fix those later.

### CODE REVIEW

**Comment by kronbichler** [comment]
Could someone take a look here? While I'm confident that all things work as expected and more efficiently than before, I'd feel safer if we have a month with this feature in deal.II before the release so some users get to pick it up.

File: source/lac/sparsity_tools.cc [hunk 1]
```diff
@@ -1040,42 +1032,51 @@ namespace SparsityTools
...
+    distribute_sparsity_pattern(dsp, owned, mpi_comm, myrange);
+  }
+
+
+
+  void
+  distribute_sparsity_pattern(DynamicSparsityPattern &dsp,
+                              const IndexSet &        owned,
+                              const MPI_Comm &        mpi_comm,
+                              const IndexSet &        myrange)
```
Comment (drwells) [inline source/lac/sparsity_tools.cc:1044]: Would you mind changing these variable names to match the ones in the declaration? It makes the code a bit easier to read.

File: source/lac/sparsity_tools.cc [hunk 2]
```diff
@@ -1099,66 +1100,67 @@ namespace SparsityTools
...
+                                myrange);
+  }
+
+
 
+  void
+  distribute_sparsity_pattern(BlockDynamicSparsityPattern &dsp,
+                              const IndexSet &             owned,
+                              const MPI_Comm &             mpi_comm,
+                              const IndexSet &             myrange)
```
Comment (drwells) [inline source/lac/sparsity_tools.cc:1124]: same

File: source/lac/sparsity_tools.cc (hunk 1 above)
Comment (kronbichler) [reply source/lac/sparsity_tools.cc:1044]: Definitely, I will do it now!

Comment (kronbichler) [reply source/lac/sparsity_tools.cc]: Done!

**Comment by kronbichler** [comment]
Any further comments?

**Comment by drwells** [comment]
Nope, lets merge!

**Pull request merged**